onnxruntime/core/providers/cpu/nn/conv_integer.cc

```diff
@@ -0,0 +1,153 @@
+  size_t num_inputs = OpKernel::Node().InputDefs().size();
+  const Tensor* X = context->Input<Tensor>(0);
+  const Tensor* W = context->Input<Tensor>(1);
+  const Tensor* X_Zero_Point = nullptr;
+  const Tensor* W_Zero_Point = nullptr;
+  int32_t input_offset = 0, filter_offset = 0;
+  if (num_inputs >= 3) {
+    X_Zero_Point = context->Input<Tensor>(2);
+    if (X_Zero_Point->Shape().NumDimensions() == 0 ||
+        (X_Zero_Point->Shape().NumDimensions() == 1 && X_Zero_Point->Shape().GetDims().size() == 1)) {
+      input_offset = static_cast<int32_t>(*(X_Zero_Point->Data<uint8_t>()));
+    } else {
+      //TODO: NOT supported.
+    }
+  }
+  if (num_inputs >= 4) {
+    W_Zero_Point = context->Input<Tensor>(3);
+    if (W_Zero_Point->Shape().NumDimensions() == 0 ||
+        (W_Zero_Point->Shape().NumDimensions() == 1 && W_Zero_Point->Shape().GetDims().size() == 1)) {
+      filter_offset = static_cast<int32_t>(*(W_Zero_Point->Data<uint8_t>()));
+    } else {
+      //TODO: NOT supported.
+    }
+  }
+
+  const int64_t N = X->Shape()[0];
+  const int64_t C = X->Shape()[1];
+  const int64_t M = W->Shape()[0];
+  ORT_RETURN_IF_ERROR(ValidateInputShape(X, W));
+
+  std::vector<int64_t> kernel_shape = ComputeKernelShape(W->Shape());
+
+  if (kernel_shape.size() + 2 != W->Shape().NumDimensions()) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "kernel_shape num_dims is not compatible with W num_dims.",
+                           " kernel_shape: ", TensorShape(kernel_shape).ToString().c_str(),
+                           " W: ", W->Shape().ToString().c_str());
+  }
+
+  for (size_t i = 0; i < kernel_shape.size(); ++i) {
+    if (kernel_shape[i] != W->Shape()[i + 2]) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "kernel_shape is not compatible with W shape.",
+                             " kernel_shape: ", TensorShape(kernel_shape).ToString().c_str(),
+                             " W: ", W->Shape().ToString().c_str());
+    }
+  }
+
+  std::vector<int64_t> pads(pads_);
+  if (pads.empty()) {
+    pads.resize(kernel_shape.size() * 2, 0);
+  }
+  std::vector<int64_t> dilations(dilations_);
+  if (dilations.empty()) {
+    dilations.resize(kernel_shape.size(), 1);
+  }
+  std::vector<int64_t> strides(strides_);
+  if (strides.empty()) {
+    strides.resize(kernel_shape.size(), 1);
+  }
+
+  std::vector<int64_t> Y_dims;
+  Y_dims.insert(Y_dims.begin(), {N, M});
+  TensorShape input_shape = X->Shape().Slice(2);
+  ORT_RETURN_IF_ERROR(InferOutputShape(input_shape, kernel_shape, strides, dilations, &pads, &Y_dims));
+  Tensor* Y = context->Output(0, TensorShape(Y_dims));
+  TensorShape output_shape = Y->Shape().Slice(2);
+
+  AllocatorPtr alloc;
+  ORT_RETURN_IF_ERROR(context->GetTempSpaceAllocator(&alloc));
+
+  const uint8_t* Xdata = X->template Data<uint8_t>();
+  int32_t* Ydata = Y->template MutableData<int32_t>();
+
+  const size_t kernel_rank = kernel_shape.size();
+
+  const int64_t input_image_size = input_shape.Size();
```

Comment:
not in scope for this review :
since there are a few commonalities in ConvInteger and Conv (and possibly QLinearConv in future) eg. intermediate calculations from 86 to 93 and more. So it may be worth to extract this in 1 palce... may be even ConvBase